File: script-worker/src/handler.js
// Lambda entrypoint. Two operations, dispatched on `event.op`:
//
//   op: 'analyze' (default — the shape core has always sent)
//     download tarball from S3 → extract → parse package.json → per-file
//     deobfuscate + AST analyze → emit findings JSON + bundle back to S3.
//     Event:    { s3Bucket, s3InputKey, projectId }
//     Response: { findingsKey, bundleKey, summary }
//
//   op: 'deobfuscate' (on-demand, analyst pressed the button on one file)
//     pull that one file out of the analysis bundle already in S3, run the
//     requested engine (or auto-select), return the source inline. No S3
//     write and no findings mutation — this is a read-only view transform,
//     so re-running it with a different engine is free and idempotent.
//     Event:    { s3Bucket, projectId, filePath, engine, bundleKey? }
//     Response: { engine, used, source, note, attempts, truncated }
//
// Errors surface as `{ error: '...' }` with non-2xx semantics in the
// Lambda invocation result (caller decides how to treat).

const fs = require('node:fs');
const fsp = require('node:fs/promises');
const path = require('node:path');
const os = require('node:os');
const tar = require('tar');

const s3io = require('./s3-io');
const archive = require('./extract');
const pkgParser = require('./package-json-parser');
const { tryDeobfuscate } = require('./deobfuscator');
const { deobfuscateOnDemand, ENGINE_IDS } = require('./engines');
const { analyzeSource } = require('./analyzer');

// 10 MB per file. Real malicious payloads run huge — the Red Hat
// preinstall droppers are 4 MB single-line `eval(...)` calls with a
// 60k-element decode array. A cap below that silently hides the worst
// findings, which is worse than burning a few extra CPU seconds.
const MAX_FILE_BYTES = Number(process.env.MAX_FILE_BYTES || 10 * 1024 * 1024);
const MAX_FILES = Number(process.env.MAX_FILES || 2000);
// AST analysis itself has a separate, lower ceiling — Babel's parser is
// O(n²) on adversarially-large single-line files. Above this we fall back
// to a regex pre-scan that still catches the obvious smells (eval,
// new Function, known-c2 hosts) without burning the full Lambda timeout.
const MAX_AST_BYTES = Number(process.env.MAX_AST_BYTES || 512 * 1024);
const ANALYZED_EXTENSIONS = new Set(['.js', '.mjs', '.cjs', '.ts']);

exports.handler = async (event, context) => {
  if (event && event.op === 'deobfuscate') {
    return deobfuscateFile(event);
  }
  return analyzePackage(event);
};

async function analyzePackage(event) {
  const startedAt = Date.now();
  const { s3Bucket, s3InputKey, projectId } = event || {};
  if (!s3Bucket || !s3InputKey || !projectId) {
    throw new Error('event missing s3Bucket / s3InputKey / projectId');
  }

  const workRoot = await fsp.mkdtemp(path.join(os.tmpdir(), `script-${projectId}-`));
  const tarballPath = path.join(workRoot, 'input.tgz');
  const extractDir = path.join(workRoot, 'extract');
  const deobfDir = path.join(workRoot, 'deobfuscated');
  const bundlePath = path.join(workRoot, 'bundle.tgz');
  const findingsKey = `scripts/${projectId}/findings.json`;
  const bundleKey = `scripts/${projectId}/deobfuscated.tgz`;

  try {
    await s3io.downloadToFile(s3Bucket, s3InputKey, tarballPath);
    const { entryCount, format } = await archive.extract(tarballPath, extractDir);
    console.log(`extracted format=${format} entries=${entryCount}`);

    const pkgInfo = pkgParser.parse(extractDir);
    const findings = pkgParser.installHookFindings(pkgInfo);

    // Walk from the package root (if we found one) so file paths in
    // findings are relative to the package, not the extraction wrapper.
    // Falls back to the extract dir when there's no package.json
    // (single-.js upload, or unparseable manifest).
    const analysisRoot = pkgInfo.packageRoot || extractDir;
    const jsFiles = walkAnalyzable(analysisRoot);
    const deobfFiles = []; // track for the bundle manifest

    for (const file of jsFiles.slice(0, MAX_FILES)) {
      const rel = path.relative(analysisRoot, file);
      const stat = await fsp.stat(file);
      if (stat.size > MAX_FILE_BYTES) {
        // Don't fail silently — a hostile 50 MB single-line file is a
        // strong negative signal on its own. Emit a finding so the user
        // sees something rather than a clean bill of health.
        findings.push(skipFinding(rel, stat.size));
        continue;
      }
      const raw = await fsp.readFile(file, 'utf8');
      const { source: deobfSource, used: deobfUsed } = await tryDeobfuscate(raw);
      if (deobfUsed) {
        const outPath = path.join(deobfDir, rel);
        await fsp.mkdir(path.dirname(outPath), { recursive: true });
        await fsp.writeFile(outPath, deobfSource, 'utf8');
        deobfFiles.push(rel);
      }
      // Babel struggles with adversarially-large single-line files (the
      // Red Hat preinstall droppers are 4 MB on a single line). For those
      // we still want SOMETHING — a regex pre-scan catches the obvious
      // dangerous primitives even without a full AST.
      if (deobfSource.length > MAX_AST_BYTES) {
        findings.push(...regexScan(deobfSource, rel, deobfUsed));
      } else {
        findings.push(...analyzeSource(deobfSource, rel, deobfUsed));
      }
    }

    // Escalate install-hook severity when the hook targets a file that was
    // (a) deobfuscated, or (b) carries an entropy-blob finding — either
    // condition is a strong "obfuscated install hook" signal.
    upgradeInstallHookSeverity(findings, deobfFiles);

    // Assign deterministic IDs after all rules have fired.
    findings.sort((a, b) =>
      severityRank(a.severity) - severityRank(b.severity) ||
      a.file.localeCompare(b.file) || a.line - b.line);
    findings.forEach((f, i) => { f.id = `f-${String(i + 1).padStart(4, '0')}`; });

    const findingsJson = {
      schemaVersion: 1,
      analyzedAt: new Date().toISOString(),
      durationMs: Date.now() - startedAt,
      summary: {
        fileCount: jsFiles.length,
        tarballEntryCount: entryCount,
        findingCount: findings.length,
        countsBySeverity: countsBySeverity(findings),
        package: packageSummary(pkgInfo),
        deobfuscatedFileCount: deobfFiles.length,
        ecosystem: 'npm',
      },
      findings,
    };

    await s3io.uploadJson(s3Bucket, findingsKey, findingsJson);
    // Bundle from the package root (when we found one) so the source
    // browser sees `index.js` directly instead of `tmp/tmpXXX/.../package/index.js`.
    await packBundle(analysisRoot, deobfDir, deobfFiles, bundlePath);
    await s3io.uploadFile(s3Bucket, bundleKey, bundlePath, 'application/gzip');

    return {
      findingsKey,
      bundleKey,
      summary: findingsJson.summary,
    };
  } finally {
    // Best-effort scratch cleanup — Lambda /tmp is per-container and small.
    try { await fsp.rm(workRoot, { recursive: true, force: true }); } catch (_) {}
  }
}

// --- on-demand single-file deobfuscation ---------------------------------

// Lambda's synchronous invoke response is hard-capped at 6 MB — and that
// limit applies to the SERIALIZED payload, not the raw source. JSON
// escaping inflates minified/obfuscated JS substantially (every quote,
// backslash and newline doubles; non-ASCII becomes \uXXXX), so a 4 MB
// source can serialize past 6 MB and the invoke fails with an opaque
// "Response payload size exceeded" that loses the whole result. We
// therefore budget against the encoded size and shrink until it fits.
const MAX_RESPONSE_BYTES = Number(process.env.MAX_RESPONSE_BYTES || 5 * 1024 * 1024);

/**
 * Shrink `source` until the whole response fits Lambda's payload limit.
 *
 * Measures the SERIALIZED size (see MAX_RESPONSE_BYTES) and halves the
 * source until it fits, rather than assuming a fixed expansion ratio —
 * escaping cost varies from ~1x for plain ASCII to ~6x for source full of
 * quotes and non-ASCII, which is exactly what obfuscated payloads look
 * like. Truncating beats failing: a analyst can still read the top of a
 * decoded dropper, and `truncated` tells the UI to say so.
 */
function fitToResponseLimit(response) {
  let out = response;
  let encoded = Buffer.byteLength(JSON.stringify(out), 'utf8');
  if (encoded <= MAX_RESPONSE_BYTES) return out;

  let source = out.source || '';
  // Start from a ratio-informed guess, then halve until it fits. Bounded
  // by construction: each pass at least halves the source.
  let budget = Math.floor(source.length * (MAX_RESPONSE_BYTES / encoded) * 0.9);
  while (budget > 1024) {
    source = out.source.slice(0, budget);
    const candidate = { ...out, source, truncated: true };
    encoded = Buffer.byteLength(JSON.stringify(candidate), 'utf8');
    if (encoded <= MAX_RESPONSE_BYTES) return candidate;
    budget = Math.floor(budget / 2);
  }
  // Pathological: even a tiny slice doesn't fit. Drop the source entirely
  // rather than lose the note + attempts the analyst can still act on.
  return {
    ...out,
    source: '',
    truncated: true,
    note: `${out.note} (result too large to return — ${(out.source || '').length} chars)`,
  };
}

/**
 * Deobfuscate ONE file from a project's existing analysis bundle.
 *
 * Reads from the bundle rather than accepting source in the event: the
 * bundle is the authoritative copy of what was analyzed, it keeps the
 * request payload tiny regardless of file size, and it means a caller
 * cannot ask us to burn Lambda CPU on arbitrary attacker-supplied text.
 */
async function deobfuscateFile(event) {
  const startedAt = Date.now();
  const { s3Bucket, projectId, filePath, engine, bundleKey } = event || {};
  if (!s3Bucket || !projectId || !filePath) {
    throw new Error('event missing s3Bucket / projectId / filePath');
  }
  const requestedEngine = engine || 'auto';
  if (requestedEngine !== 'auto' && !ENGINE_IDS.includes(requestedEngine)) {
    throw new Error(`unknown engine '${requestedEngine}'`);
  }

  const workRoot = await fsp.mkdtemp(path.join(os.tmpdir(), `deobf-${projectId}-`));
  const bundlePath = path.join(workRoot, 'bundle.tgz');
  const outDir = path.join(workRoot, 'out');
  try {
    // Caller passes the bundle key recorded on the analysis row; the
    // conventional path is only a fallback for older rows.
    const key = bundleKey || `scripts/${projectId}/deobfuscated.tgz`;
    await s3io.downloadToFile(s3Bucket, key, bundlePath);

    // Extract only the one entry we need. `original/` is the untouched
    // upload tree — always deobfuscate from that, never from a previously
    // deobfuscated copy, so engines are compared on equal footing.
    const wanted = `original/${filePath}`;
    await fsp.mkdir(outDir, { recursive: true });
    await tar.x({ file: bundlePath, cwd: outDir, filter: (p) => p === wanted || p === `./${wanted}` });

    const onDisk = path.join(outDir, wanted);
    let raw;
    try {
      raw = await fsp.readFile(onDisk, 'utf8');
    } catch (_) {
      throw new Error(`file not found in bundle: ${filePath}`);
    }

    const result = deobfuscateOnDemand(raw, requestedEngine);
    const response = fitToResponseLimit({
      ...result,
      truncated: false,
      durationMs: Date.now() - startedAt,
    });
    console.log(
      `deobfuscate project=${projectId} file=${filePath} requested=${requestedEngine} ` +
      `picked=${result.engine} used=${result.used} truncated=${response.truncated} ` +
      `ms=${Date.now() - startedAt}`,
    );
    return response;
  } finally {
    try { await fsp.rm(workRoot, { recursive: true, force: true }); } catch (_) {}
  }
}

// -----------------------------------------------------------------------

// Emitted when a file is too large to open for analysis. The size
// threshold is conservative; legitimate libraries don't ship multi-MB
// single-line JS, so passing it is itself suspicious.
function skipFinding(file, sizeBytes) {
  return {
    rule: 'oversized-file',
    severity: 'HIGH',
    file,
    line: 0,
    column: 0,
    message: `File ${file} is ${(sizeBytes / 1024 / 1024).toFixed(1)} MB — too large to analyze in full. ` +
             `Real-world malicious droppers often hide encoded payloads in oversized single-line files; manual review required.`,
    snippet: '',
    remediation: 'Open the file by hand and look for eval(), new Function(), or numeric-array decoders that reconstruct a hidden payload.',
    evidence: { sizeBytes },
    deobfuscated: false,
  };
}

// Regex pre-scan used when Babel can't (or shouldn't) parse the source —
// adversarially-large files, syntax errors that errorRecovery can't
// salvage. Covers the most decisive subset of rules: eval-surface,
// secret-theft (env reads of credential names), known-c2 string hits,
// child_process invocations. Less precise than the AST pass but never
// returns silence on a hostile input.
const RX_EVAL = /\beval\s*\(/g;
const RX_NEW_FUNCTION = /\bnew\s+Function\s*\(/g;
const RX_PROCESS_ENV = /\bprocess\s*\.\s*env\s*\.\s*([A-Z][A-Z0-9_]*)/g;
const RX_CHILD_PROC = /\b(child_process\.)?(spawn|execSync|exec|execFile|fork)\s*\(/g;
const RX_REQUIRE_CP = /\brequire\s*\(\s*['"]child_process['"]\s*\)/g;
const RX_FETCH = /\bfetch\s*\(\s*['"`]([^'"`]+)['"`]/g;

const SENSITIVE_ENV = /^(AWS_|NPM_TOKEN|GH_TOKEN|GITHUB_TOKEN|DOCKER_|NODE_AUTH_TOKEN|GCP_|AZURE_|VAULT_|SLACK_TOKEN|STRIPE_|TWILIO_|SENDGRID_)/;
const KNOWN_C2 = /(discord\.com\/api\/webhooks|webhook\.site|requestbin\.(net|com)|pastebin\.com\/raw|transfer\.sh|t\.me\/[A-Za-z0-9_]+|ipinfo\.io|npmjs\.help|workers\.dev\/[a-z0-9-]+\/(api|exfil|drop))/i;

function regexScan(source, file, deobfFlag) {
  const findings = [];
  const lineOffsets = [0];
  for (let i = 0; i < source.length; i++) if (source[i] === '\n') lineOffsets.push(i + 1);
  const lineOf = (idx) => {
    let lo = 0, hi = lineOffsets.length - 1;
    while (lo <= hi) {
      const mid = (lo + hi) >> 1;
      if (lineOffsets[mid] <= idx) lo = mid + 1; else hi = mid - 1;
    }
    return hi + 1;
  };
  const snippetAt = (idx, len) => source.slice(idx, idx + len).replace(/\s+/g, ' ').slice(0, 160);
  const push = (rule, severity, idx, message, evidence) => findings.push({
    rule, severity, file, line: lineOf(idx), column: 0, message,
    snippet: snippetAt(idx, 80), remediation: '', evidence: evidence || {}, deobfuscated: deobfFlag,
  });

  let m;
  RX_EVAL.lastIndex = 0;
  if ((m = RX_EVAL.exec(source))) push('eval-surface', 'HIGH', m.index, 'Direct eval() call detected via regex pre-scan (file too large for full AST analysis)', { kind: 'eval' });
  RX_NEW_FUNCTION.lastIndex = 0;
  if ((m = RX_NEW_FUNCTION.exec(source))) push('eval-surface', 'HIGH', m.index, 'new Function(...) detected via regex pre-scan', { kind: 'new-function' });
  RX_REQUIRE_CP.lastIndex = 0;
  if ((m = RX_REQUIRE_CP.exec(source))) push('spawn', 'HIGH', m.index, 'require("child_process") detected via regex pre-scan', { kind: 'require-child-process' });
  RX_CHILD_PROC.lastIndex = 0;
  if ((m = RX_CHILD_PROC.exec(source))) push('spawn', 'HIGH', m.index, `${m[0]} detected via regex pre-scan`, { call: m[0].trim() });

  RX_PROCESS_ENV.lastIndex = 0;
  const seenEnv = new Set();
  while ((m = RX_PROCESS_ENV.exec(source)) && seenEnv.size < 10) {
    if (!SENSITIVE_ENV.test(m[1])) continue;
    if (seenEnv.has(m[1])) continue;
    seenEnv.add(m[1]);
    push('secret-theft', 'CRITICAL', m.index, `Reads sensitive env var ${m[1]} (regex pre-scan)`, { envVar: m[1] });
  }

  RX_FETCH.lastIndex = 0;
  while ((m = RX_FETCH.exec(source))) {
    const url = m[1];
    push('net-exfil', 'HIGH', m.index, `fetch() to ${url} (regex pre-scan)`, { call: 'fetch', target: url });
    if (KNOWN_C2.test(url)) {
      push('known-c2', 'CRITICAL', m.index, `URL ${url} matches a known-bad indicator`, { indicator: url });
    }
  }

  return findings;
}

function walkAnalyzable(root) {
  const out = [];
  const stack = [root];
  while (stack.length) {
    const dir = stack.pop();
    let entries;
    try { entries = fs.readdirSync(dir, { withFileTypes: true }); }
    catch (_) { continue; }
    for (const e of entries) {
      const p = path.join(dir, e.name);
      if (e.isDirectory()) {
        if (e.name === 'node_modules' || e.name === '.git') continue;
        stack.push(p);
      } else if (e.isFile()) {
        const ext = path.extname(e.name).toLowerCase();
        if (ANALYZED_EXTENSIONS.has(ext)) out.push(p);
      }
    }
  }
  return out;
}

function countsBySeverity(findings) {
  const c = { CRITICAL: 0, HIGH: 0, MEDIUM: 0, INFO: 0 };
  for (const f of findings) {
    if (c[f.severity] != null) c[f.severity] += 1;
  }
  return c;
}

function packageSummary(pkgInfo) {
  if (!pkgInfo.found) return { found: false };
  return {
    found: true,
    name: pkgInfo.name,
    version: pkgInfo.version,
    description: pkgInfo.description,
    maintainerCount: pkgInfo.maintainers.length,
    dependencyCount: pkgInfo.dependencyCount,
    hasInstallHook: pkgInfo.hasInstallHook,
    installHooks: pkgInfo.hooks,
    parseError: pkgInfo.parseError || null,
  };
}

function upgradeInstallHookSeverity(findings, deobfFiles) {
  const hookFindings = findings.filter((f) => f.rule === 'install-hook');
  if (hookFindings.length === 0) return;

  // Build the "high-signal" file set: anything that carries a CRITICAL or
  // HIGH finding, plus anything the deobfuscator touched. Install-hook
  // pointed at any of these means the malicious code chain is wired up
  // through the install path — the canonical supply-chain attack shape.
  const dangerousFiles = new Set(deobfFiles);
  for (const f of findings) {
    if (f.rule === 'install-hook') continue;
    if (f.severity === 'CRITICAL' || f.severity === 'HIGH') {
      dangerousFiles.add(f.file);
    }
  }

  for (const h of hookFindings) {
    const script = h.evidence?.script || '';
    const targets = [...dangerousFiles].filter((df) => script.includes(path.basename(df)) || script.includes(df));
    if (targets.length > 0) {
      // Hook targets a file that already raised flags — this combination
      // is the textbook supply-chain attack shape, bump to CRITICAL.
      h.severity = 'CRITICAL';
      h.message += ` — target file ${targets[0]} carries high-severity findings`;
    }
  }
}

function severityRank(s) {
  return { CRITICAL: 0, HIGH: 1, MEDIUM: 2, INFO: 3 }[s] ?? 4;
}

async function packBundle(extractDir, deobfDir, deobfFiles, outPath) {
  const manifest = {
    schemaVersion: 1,
    deobfuscatedFiles: deobfFiles,
  };
  // Write manifest into a temp dir alongside the two trees.
  const stageDir = path.join(path.dirname(outPath), 'stage');
  const originalOut = path.join(stageDir, 'original');
  const deobfOut = path.join(stageDir, 'deobfuscated');
  await fsp.mkdir(stageDir, { recursive: true });
  await copyDir(extractDir, originalOut);
  if (deobfFiles.length > 0) {
    await copyDir(deobfDir, deobfOut);
  }
  await fsp.writeFile(path.join(stageDir, 'manifest.json'), JSON.stringify(manifest, null, 2));
  await tar.c(
    { gzip: true, file: outPath, cwd: stageDir },
    ['manifest.json', 'original', ...(deobfFiles.length > 0 ? ['deobfuscated'] : [])],
  );
}

async function copyDir(src, dest) {
  await fsp.mkdir(dest, { recursive: true });
  const entries = await fsp.readdir(src, { withFileTypes: true });
  for (const e of entries) {
    const sp = path.join(src, e.name);
    const dp = path.join(dest, e.name);
    if (e.isDirectory()) await copyDir(sp, dp);
    else if (e.isFile()) await fsp.copyFile(sp, dp);
  }
}

// Internals exposed for tests only — the Lambda contract is `handler`.
exports.__test__ = { fitToResponseLimit, MAX_RESPONSE_BYTES };

File: pypi-worker/src/handler.py
"""Lambda entrypoint. Orchestrates:
  download upload from S3 → extract → parse setup.py/pyproject/METADATA
  → per-.py AST analyze → emit findings JSON + bundle back to S3.

Event shape (identical to script-worker for Spring-side parity):
  { s3Bucket, s3InputKey, projectId }

Response shape:
  { findingsKey, bundleKey, summary }
"""

import os
import re
import shutil
import sys
import tarfile
import tempfile
import time
import traceback

import s3_io
import extractor
import manifest_parser
from analyzer import analyze_source


MAX_FILE_BYTES = int(os.environ.get("MAX_FILE_BYTES", str(10 * 1024 * 1024)))
MAX_FILES = int(os.environ.get("MAX_FILES", "2000"))
MAX_AST_BYTES = int(os.environ.get("MAX_AST_BYTES", str(512 * 1024)))
ANALYZED_EXTENSIONS = {".py"}


def handler(event, context):
    started_at = time.time()
    s3_bucket = (event or {}).get("s3Bucket")
    s3_input_key = (event or {}).get("s3InputKey")
    project_id = (event or {}).get("projectId")
    if not s3_bucket or not s3_input_key or not project_id:
        raise RuntimeError("event missing s3Bucket / s3InputKey / projectId")

    work_root = tempfile.mkdtemp(prefix=f"pypi-{project_id}-")
    input_path = os.path.join(work_root, "input.bin")
    extract_dir = os.path.join(work_root, "extract")
    bundle_path = os.path.join(work_root, "bundle.tgz")
    findings_key = f"scripts/{project_id}/findings.json"
    bundle_key = f"scripts/{project_id}/deobfuscated.tgz"

    try:
        s3_io.download_to_file(s3_bucket, s3_input_key, input_path)
        try:
            ext_meta = extractor.extract(input_path, extract_dir)
        except Exception as e:
            # Surface zip/tar problems as a worker error so Spring can map
            # the message to a friendlier 400 (Datadog encrypted dataset).
            raise RuntimeError(str(e))
        print(f"extracted format={ext_meta['format']} entries={ext_meta['entryCount']}", file=sys.stderr)

        pkg_info = manifest_parser.parse(extract_dir)
        findings = manifest_parser.install_hook_findings(pkg_info)

        analysis_root = pkg_info.get("packageRoot") or extract_dir
        py_files = _walk_analyzable(analysis_root)

        for file in py_files[:MAX_FILES]:
            rel = os.path.relpath(file, analysis_root)
            try:
                stat = os.stat(file)
            except OSError:
                continue
            if stat.st_size > MAX_FILE_BYTES:
                findings.append(_skip_finding(rel, stat.st_size))
                continue
            try:
                with open(file, "r", encoding="utf-8", errors="replace") as f:
                    source = f.read()
            except OSError:
                continue
            if len(source) > MAX_AST_BYTES:
                findings.extend(_regex_scan(source, rel))
            else:
                findings.extend(analyze_source(source, rel, deobf_flag=False))

        _upgrade_install_hook_severity(findings)

        findings.sort(key=lambda f: (_sev_rank(f["severity"]), f["file"], f["line"]))
        for i, f in enumerate(findings):
            f["id"] = f"f-{i+1:04d}"

        findings_json = {
            "schemaVersion": 1,
            "analyzedAt": _iso_now(),
            "durationMs": int((time.time() - started_at) * 1000),
            "summary": {
                "fileCount": len(py_files),
                "tarballEntryCount": ext_meta["entryCount"],
                "findingCount": len(findings),
                "countsBySeverity": _counts_by_severity(findings),
                "package": _package_summary(pkg_info),
                "deobfuscatedFileCount": 0,
                "ecosystem": "pypi",
            },
            "findings": findings,
        }

        s3_io.upload_json(s3_bucket, findings_key, findings_json)
        _pack_bundle(analysis_root, bundle_path)
        s3_io.upload_file(s3_bucket, bundle_key, bundle_path, "application/gzip")

        return {
            "findingsKey": findings_key,
            "bundleKey": bundle_key,
            "summary": findings_json["summary"],
        }
    except Exception:
        traceback.print_exc()
        raise
    finally:
        try:
            shutil.rmtree(work_root, ignore_errors=True)
        except Exception:
            pass


# -----------------------------------------------------------------------


def _walk_analyzable(root: str) -> list:
    out = []
    for dirpath, dirnames, filenames in os.walk(root):
        dirnames[:] = [d for d in dirnames if d not in {
            "__pycache__", ".git", "node_modules", ".tox", ".venv", "venv",
        }]
        for name in filenames:
            if os.path.splitext(name)[1].lower() in ANALYZED_EXTENSIONS:
                out.append(os.path.join(dirpath, name))
    return out


def _skip_finding(file: str, size_bytes: int) -> dict:
    return {
        "rule": "oversized-file",
        "severity": "HIGH",
        "file": file,
        "line": 0,
        "column": 0,
        "message": (
            f"File {file} is {size_bytes / 1024 / 1024:.1f} MB — too large to analyze in full. "
            f"Hostile Python files at this size are typically a single base64 blob feeding exec(); manual review required."
        ),
        "snippet": "",
        "remediation": "Open the file by hand; look for exec(), base64.b64decode + exec, or marshal.loads.",
        "evidence": {"sizeBytes": size_bytes},
        "deobfuscated": False,
    }


_RX_EVAL = re.compile(r"\beval\s*\(")
_RX_EXEC = re.compile(r"\bexec\s*\(")
_RX_SUBPROC = re.compile(r"\b(subprocess|os)\.(system|popen|run|Popen|call|check_call|check_output|getoutput)\s*\(")
_RX_OS_ENV = re.compile(r"os\.environ(?:\.get)?\s*[\(\[]\s*['\"]([A-Z][A-Z0-9_]*)['\"]")
_RX_GETENV = re.compile(r"os\.getenv\s*\(\s*['\"]([A-Z][A-Z0-9_]*)['\"]")
_RX_URLLIB = re.compile(r"(urllib\.request\.urlopen|requests\.(get|post|put)|httpx\.(get|post))\s*\(\s*['\"]([^'\"]+)['\"]")

_SENSITIVE_ENV_RX = re.compile(r"^(AWS_|NPM_TOKEN|GH_TOKEN|GITHUB_TOKEN|DOCKER_|NODE_AUTH_TOKEN|GCP_|AZURE_|VAULT_|SLACK_TOKEN|STRIPE_|TWILIO_|SENDGRID_|PYPI_|TWINE_|POETRY_)")
_KNOWN_C2_RX = re.compile(r"(discord\.com/api/webhooks|webhook\.site|requestbin\.(net|com)|pastebin\.com/raw|transfer\.sh|t\.me/[A-Za-z0-9_]+|ipinfo\.io|workers\.dev/[a-z0-9-]+/(api|exfil|drop))", re.I)


def _regex_scan(source: str, file: str) -> list:
    """Fallback for files Babel-... err, ast.parse can't handle in time.
    Less precise than the AST pass but never returns silence on hostile
    input."""
    findings = []
    line_offsets = [0]
    for i, c in enumerate(source):
        if c == "\n":
            line_offsets.append(i + 1)

    def line_of(idx: int) -> int:
        lo, hi = 0, len(line_offsets) - 1
        while lo <= hi:
            mid = (lo + hi) // 2
            if line_offsets[mid] <= idx:
                lo = mid + 1
            else:
                hi = mid - 1
        return hi + 1

    def push(rule: str, severity: str, idx: int, message: str, evidence: dict | None = None) -> None:
        findings.append({
            "rule": rule, "severity": severity, "file": file,
            "line": line_of(idx), "column": 0, "message": message,
            "snippet": source[idx:idx+80].replace("\n", " ")[:160],
            "remediation": "", "evidence": evidence or {}, "deobfuscated": False,
        })

    if (m := _RX_EVAL.search(source)):
        push("eval-surface", "HIGH", m.start(), "eval() detected via regex pre-scan (file too large for AST)", {"kind": "eval"})
    if (m := _RX_EXEC.search(source)):
        push("eval-surface", "HIGH", m.start(), "exec() detected via regex pre-scan", {"kind": "exec"})
    if (m := _RX_SUBPROC.search(source)):
        push("spawn", "HIGH", m.start(), f"{m.group(0).strip()} detected via regex pre-scan", {"call": m.group(0)})

    seen_env = set()
    for m in _RX_OS_ENV.finditer(source):
        env = m.group(1)
        if env in seen_env or not _SENSITIVE_ENV_RX.match(env):
            continue
        seen_env.add(env)
        push("secret-theft", "CRITICAL", m.start(), f"Reads sensitive env var {env} (regex pre-scan)", {"envVar": env})
        if len(seen_env) >= 10:
            break
    for m in _RX_GETENV.finditer(source):
        env = m.group(1)
        if env in seen_env or not _SENSITIVE_ENV_RX.match(env):
            continue
        seen_env.add(env)
        push("secret-theft", "CRITICAL", m.start(), f"Reads sensitive env var {env} (regex pre-scan)", {"envVar": env})

    for m in _RX_URLLIB.finditer(source):
        target = m.group(4)
        push("net-exfil", "HIGH", m.start(), f"{m.group(1)}() to {target} (regex pre-scan)", {"call": m.group(1), "target": target})
        if _KNOWN_C2_RX.search(target):
            push("known-c2", "CRITICAL", m.start(), f"URL {target} matches a known-bad indicator", {"indicator": target})

    return findings


def _counts_by_severity(findings: list) -> dict:
    counts = {"CRITICAL": 0, "HIGH": 0, "MEDIUM": 0, "INFO": 0}
    for f in findings:
        sev = f.get("severity")
        if sev in counts:
            counts[sev] += 1
    return counts


def _package_summary(info: dict) -> dict:
    if not info.get("found"):
        return {"found": False}
    return {
        "found": True,
        "name": info.get("name"),
        "version": info.get("version"),
        "description": info.get("description"),
        "maintainerCount": len(info.get("maintainers") or []),
        "dependencyCount": info.get("dependencyCount", 0),
        "hasInstallHook": info.get("hasInstallHook", False),
        "installHooks": [
            {"key": h.get("key"), "script": h.get("script", "")}
            for h in (info.get("hooks") or [])
        ],
        "parseError": info.get("parseError"),
    }


def _upgrade_install_hook_severity(findings: list) -> None:
    """Mirror of the NPM-side escalation: if an install-hook references a
    file that already raises CRITICAL/HIGH findings, the chain is wired up
    — bump the hook itself to CRITICAL."""
    hooks = [f for f in findings if f["rule"] == "install-hook"]
    if not hooks:
        return
    dangerous_files = set()
    for f in findings:
        if f["rule"] == "install-hook":
            continue
        if f["severity"] in {"CRITICAL", "HIGH"}:
            dangerous_files.add(f["file"])
    for h in hooks:
        script = (h.get("evidence") or {}).get("script") or ""
        targets = [df for df in dangerous_files if df in script or os.path.basename(df) in script]
        if targets:
            h["severity"] = "CRITICAL"
            h["message"] += f" — referenced file {targets[0]} carries high-severity findings"


def _sev_rank(s: str) -> int:
    return {"CRITICAL": 0, "HIGH": 1, "MEDIUM": 2, "INFO": 3}.get(s, 4)


def _pack_bundle(analysis_root: str, out_path: str) -> None:
    """Pack the analyzed tree into a tar.gz in the same shape the frontend
    bundle viewer already understands: `manifest.json` + `original/` tree.
    No `deobfuscated/` for PyPI yet — no Python deobfuscator wired."""
    stage = os.path.join(os.path.dirname(out_path), "stage")
    original = os.path.join(stage, "original")
    os.makedirs(original, exist_ok=True)
    _copy_tree(analysis_root, original)
    manifest = {"schemaVersion": 1, "deobfuscatedFiles": []}
    import json
    with open(os.path.join(stage, "manifest.json"), "w") as f:
        json.dump(manifest, f, indent=2)
    with tarfile.open(out_path, "w:gz") as tar:
        tar.add(os.path.join(stage, "manifest.json"), arcname="manifest.json")
        tar.add(original, arcname="original")


def _copy_tree(src: str, dst: str) -> None:
    for dirpath, dirnames, filenames in os.walk(src):
        rel = os.path.relpath(dirpath, src)
        target = dst if rel == "." else os.path.join(dst, rel)
        os.makedirs(target, exist_ok=True)
        for name in filenames:
            try:
                shutil.copyfile(os.path.join(dirpath, name), os.path.join(target, name))
            except OSError:
                pass


def _iso_now() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
